doc/nextml2015/presentation.tex

```diff
@@ -0,0 +1,1186 @@
+>>> updates = OrderedDict()
+>>> updates[x] = x + 1
+>>> f = function([], updates=updates)
+>>> f()
+>>> x.get_value()
+1.0
+>>> x.set_value(100.)
+>>> f()
+>>> x.get_value()
+101.0
+\end{lstlisting}
+\end{frame}
+
+\begin{frame}{Which dict?}
+  \begin{itemize}
+  \item Use theano.compat.python2x.OrderedDict
+  \item Not collections.OrderedDict
+  \begin{itemize}
+  \item This isn’t available in older versions of python
+  \end{itemize}
+  \item Not \{\} aka dict
+  \begin{itemize}
+  \item The iteration order of this built-in class is not
+    deterministic (thanks, Python!) so if Theano
+    accepted this, the same script could compile
+    different C programs each time you run it
+  \end{itemize}
+  \end{itemize}
+\end{frame}
+
+\begin{frame}{Compilation modes}
+  \begin{itemize}
+  \item Can compile in different modes to get different kinds of programs
+  \item Can specify these modes very precisely with arguments to theano.function
+  \item Can use a few quick presets with environment variable flags
+  \end{itemize}
+\end{frame}
+
+\begin{frame}{Example preset compilation modes}
+  \begin{itemize}
+  \item FAST\_RUN: default. Fastest execution, slowest compilation
+  \item FAST\_COMPILE: Fastest compilation, slowest execution. No C code.
+  \item DEBUG\_MODE: Adds lots of checks.
+Raises error messages in situations other
+modes regard as fine.
+  \item optimizer=fast\_compile: as mode=FAST\_COMPILE, but with C code.
+  \item theano.function(..., mode=``FAST\_COMPILE'')
+  \item THEANO\_FLAGS=mode=FAST\_COMPILE python script.py
+  \end{itemize}
+\end{frame}
+
+\begin{frame}{Compilation for GPU}
+  \begin{itemize}
+  \item Theano current back-end only supports 32 bit on GPU
+  \item libgpuarray (new-backend) support all dtype
+  \item CUDA supports 64 bit, but is slow in gamer card
+  \item T.fscalar, T.fvector, T.fmatrix are all 32 bit
+  \item T.scalar, T.vector, T.matrix resolve to 32 bit or 64 bit depending on theano’s floatX flag
+  \item floatX is float64 by default, set it to float32
+  \item Set device flag to gpu (or a specific gpu, like gpu0)
+  \item Flag: warn\_float64={'ignore', 'warn', 'raise', 'pdb'}
+  \end{itemize}
+\end{frame}
+
+\subsection{Modifying expressions}
+\begin{frame}{Modifying expressions}
+  \begin{itemize}
+  \item The grad method
+  \item Others
+
+%  \item Variable nodes
+%  \item Types
+%  \item Ops
+%  \item Apply nodes
+  \end{itemize}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{The grad method}
+
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+>>> x = T.scalar('x')
+>>> y = 2. * x
+>>> g = T.grad(y, x)
+# Print the not optimized graph
+>>> theano.printing.debugprint(g)
+Elemwise{mul} [@A] ''
+ |Elemwise{second,no_inplace} [@B] ''
+ | |Elemwise{mul,no_inplace} [@C] ''
+ | | |TensorConstant{2.0} [@D]
+ | | |x [@E]
+ | |TensorConstant{1.0} [@F]
+ |TensorConstant{2.0} [@D]
```

Comment:
I'm not convinced that a debugprint is the clearest way to show the graph.  However pprint is not much better.  Maybe a pydotprint?